### PR TITLE
Make hb with each obs message

### PR DESCRIPTION
**Pull request by KaraMelih**
Adding a heartbeat message to any significance, time, or coincidence message if there were already no heartbeats

### CODE REVIEW

**Comment by KaraMelih** [comment]
might need to check `is_test` argument for the extra created heartbeat. For now, it doesn't pass anything.

**Comment by KaraMelih** [comment]
Okay, the tests are failing because now they do not expect to see a heartbeat with other message type creations. The tests needs to be updated.

**Review by sybenzvi — CHANGES_REQUESTED** [review]
Questions:

- [x] It's OK for `is_valid` to return true even with the missing p value?
- [ ] Comparing `tier` to the message class names is awkward, is there a nicer way to do that?

**Comment by KaraMelih** [comment]
> It's OK for is_valid to return true even with the missing p value?

Yes, I think we said we should not enforce p values, even without that, an experiment should still be able to form a coincidence. That's why they are not in the "required fields" (also formulated like this in the paper) but the validation was (is) treating them as they were a required argument.

> Comparing tier to the message class names is awkward, is there a nicer way to do that?
Is this referring to the L504 ? Not sure if there is a nicer way, if you can think of something, I'm happy to implement it.

**Review by habig — COMMENTED** [review]
Changes make sense to me.  Did the test with this successfully populate the live-ness to avoid the crash when counting live detectors?

**Comment by KaraMelih** [comment]
I didn't test it but I expect that this would still not work. Because, even though I send a heartbeat with every message, they are being sent after the coincidence message. -Well actually, if the 2nd detector does not sent it at the same time, the heartbeat of the first will already be registered at the time of coincidence forming, but the heartbeat of the 2nd one will not yet be registered. 

In the production, this should not be a problem, as we should expect detectors to send heartbeats. In worst case, we append a wrong FAR estimation. In any case, we also added a fail-safe in FAR calculation (smt like; if heartbeat cache empty return 0).

TL;DR it won't crash, FAR might still be slightly wrong if the experiments were not sending heartbeats already.

**Comment by KaraMelih** [comment]
Together with @sybenzvi we implemented this logic on the server side, see this PR https://github.com/SNEWS2/SNEWS_Coincidence_System/pull/123 

Now instead of creating 2 messages with snews_pt and sending them separately, snews_cs registers a heartbeat each time it sees a coincidence message in the stream and **after that** it continues with the coincidence search. This also avoid the problems discussed above.

**Comment by KaraMelih** [comment]
I'll close this PR as it is no longer needed. I'll make a small patch with another PR for the p value issues this PR was addressing. (i.e. p values are not required arguments for coincidence and time tiers but they still try to validate, and crash if it is None.)
